Documentation

jhollowayj opened the pull request:
Here's some prelim documentation.  Links should mostly be working.  

Please click the `View` button to see how this will render as you review the PR.

renaissanceGeek commented on README.md:
```diff
@@ -1,2 +1,63 @@
+No changes to these files should ever be needed.
+
+[example-client.cc](src/example-client.cc) provides a template for building out your own c++ based client.
+Please note that details on the API itself are provided at the [master cobaltspeech/client repository](https://github.com/cobaltspeech/client), so we will only address items that are specific to c++ in this document.
```
Oh just saw the other PR--thank you! 🥇

jhollowayj commented on README.md:
```diff
@@ -1,2 +1,63 @@
+
+## Downloading and Building the Example Client SDK
+
+With everything set up from the [Pre-requisites](./#pre-rerequisites) section, the following should work without a flaw.
```
So, maybe we should re-organize the files to be something like `src/gen` for the actual SDK files and `src/example/client.cc` for examples?

As far as maintaining an intermediate layer, I have mixed feelings on that.  I think a sample would _probably_ be better, since there are quite a few configuration settings that are required for each call.  I think that if the example(s) demonstrated enough combinations, consumers of the SDK would likely be able to pick and choose which code snippets they want to use.   I feel like providing the intermediate layer (the `CubicClient` class defined in the example-client.cc file) would have to expose ways to set the various configurations for the function calls, and it doesn't seem to provide that much value.

With that said, we could maybe provide a few shortcut functions for building those configs.  Something like configs for `MakeRecognizeMP3Configs()` and `MakeStreamingRecognizeWAVConfigs()`.  Something like that could be nice.  🤷‍♂️ Thoughts?

(We should probably add a license to the repo that let's them know they can do that.)

jhollowayj commented on README.md:
```diff
@@ -1,2 +1,63 @@
+
+If you would like to verify things have been installed correctly, you can try going through the steps defined in their [cpp examples folder](https://github.com/grpc/grpc/tree/master/examples/cpp/).  
+If you can successfully run their [Hello World](https://github.com/grpc/grpc/tree/master/examples/cpp/helloworld) example, building our example cubic client should be straight forward.
+_Note_: One of the steps in the Hello World example requires a working `protoc` binary.  Some have had success also [building it from source](https://github.com/grpc/grpc/blob/master/BUILDING.md#protoc), while others have simply downloaded a binary from [github](https://github.com/protocolbuffers/protobuf/releases/latest).
```
I debated on this one.  
* To run grpc's example, you have to compile the .proto files into the .cc and .h files via the `protoc` command.
* To run our example client, you do _not_ have to compile the files.  We provide those for them.

So, without protoc (and the grpc_cpp_plugin), they can't run the grpc examples.

With that said, what would you like to do with the paragraph?

jhollowayj commented:
@renaissanceGeek I've updated the docs, shuffled files around, and organized the readme with a few more section headers.  

1. Would you mind giving the entire readme a fresh review?
2. I moved the protoc "SDK" files to the src/sdk folder and moved the sample client code to the src/examples folder.  Do you think that this change will be clear enough in the distinction between the two?  Do we need to add some documentation on the topic of what to include in their own C++ projects?  Something like "If you are integrating this SDK into an existing project, you only need to copy over the src/sdk folder and it's contents."
3. I modified the Makefile to keep a list of the SDK files separate from the example files.  I was thinking that would make it easier to build individual example clients if we ever had more than one.  Let me know if any part of the makefile is confusing or would benefit from further documentation.

Thanks!
